cmd/dep/root_analyzer.go

```diff
@@ -139,6 +140,26 @@ func (a *rootAnalyzer) FinalizeRootManifestAndLock(m *dep.Manifest, l *dep.Lock)
 			delete(m.Constraints, pr)
 		}
 	}
+	// Pick the direct dependencies from the solution lock and add to manifest.
+	// This is done to fill up the manifest constraints with the dependencies
+	// solved over the network.
+	for _, y := range l.Projects() {
+		var f *fb.ConstraintFeedback
+		pr := y.Ident().ProjectRoot
+		if _, ok := a.directDeps[string(pr)]; ok {
+			pp := getProjectPropertiesFromVersion(y.Version())
+			if pp.Constraint != nil {
+				m.Constraints[pr] = pp
+				pc := gps.ProjectConstraint{Ident: y.Ident(), Constraint: pp.Constraint}
+				f = fb.NewConstraintFeedback(pc, fb.DepTypeDirect)
+			} else {
+				f = fb.NewLockedProjectFeedback(y, fb.DepTypeDirect)
+			}
+		} else {
+			f = fb.NewLockedProjectFeedback(y, fb.DepTypeTransitive)
+		}
+		f.LogFeedback(a.ctx.Err)
+	}
```

Comment:
@darkowlzz  I wrote out the workflow for init to help answer this question and decided to dump my notes, instead of just the one part that addresses the question. 😁 

1. `rootAnalyzer.importManifestAndLock`
    * Import constraints and locked projects from external config.
1. `gopathScanner.InitializeManifestAndLock`
    * Add constraints and locked projects from GOPATH.
1. SOLVE
    * Uses dark magic to add more goodies to the lock.
1. `rootAnalyzer.FinalizeManifestAndLock`
    * Remove unused constraints. I'll remove this in a separate PR. I just realized that it's not needed anymore thanks to `rootAnalyzer.removeTransitiveDependencies`.
    * This PR adds printing feedback for direct deps found during solve.
    * This PR adds printing feedback for transitive deps locked during solve.
1. `gopathScanner.FinalizeManifestAndLock`
    * Add constraints found during solve that weren't initially in the manifest
      because they weren't on disk.

I believe that we can generalize the `FinalizeManifestAndLock` activities to "Identify new constraints and locked projects found during solve".  New constraints are direct deps that are in the final lock but not the manifest, and new locked projects are identified by keeping track of the initial lock. The same finalization tasks can be performed of how we got to this state:
  * `-gopath` wasn't specified and nothing was imported, 100% network solve.
  * The GOPATH didn't have repos for all of the direct deps.
  * The imported config was incomplete and didn't have constraints for all of the direct deps.

Then we could have init keep a copy of the lock given to the solver, and pass it in as an additional argument to `rootAnalyzer.FinalizeManifestAndLock`. Now all that logic is in one place and we can also remove `gopathScanner.FinalizeManifestAndLock`.

What do you think?